lib/spack/llnl/util/filesystem.py

```diff
@@ -1021,6 +1021,75 @@ def traverse_tree(source_root, dest_root, rel_path='', **kwargs):
         yield (source_path, dest_path)
 
 
+def lexists_islink_isdir(path):
+    """Computes the tuple (lexists(path), islink(path), isdir(path)) in a minimal
+    number of stat calls."""
+    # First try to lstat, so we know if it's a link or not.
+    try:
+        lst = os.lstat(path)
+    except (IOError, OSError):
+        return False, False, False
+
+    is_link = stat.S_ISLNK(lst.st_mode)
+
+    # Check whether file is a dir.
+    if not is_link:
+        is_dir = stat.S_ISDIR(lst.st_mode)
+        return True, is_link, is_dir
+
+    # Check whether symlink points to a dir.
+    try:
+        st = os.stat(path)
+        is_dir = stat.S_ISDIR(st.st_mode)
+    except (IOError, OSError):
+        # Dangling symlink (i.e. it lexists but not exists)
+        is_dir = False
+
+    return True, is_link, is_dir
+
+
+def visit_directory_tree(root, visitor, rel_path='', depth=0):
+    """
+    Recurses the directory root depth-first through a visitor pattern
+
+    The visitor interface is as follows:
+    - visit_file(root, rel_path, depth)
+    - before_visit_dir(root, rel_path, depth) -> bool
+        if True, descends into this directory
+    - before_visit_symlinked_dir(root, rel_path, depth) -> bool
+        if True, descends into this directory
+    - after_visit_dir(root, rel_path, depth) -> void
+        only called when before_visit_dir returns True
+    - after_visit_symlinked_dir(root, rel_path, depth) -> void
+        only called when before_visit_symlinked_dir returns True
+    """
+    dir = os.path.join(root, rel_path)
+
+    dir_entries = os.listdir(dir)
```

Comment:
Ok, here I have questions.  This function looks a good bit like `os.walk()`, except that it does its own stating.  If the goal is performance, strongly consider `os.scandir` for this, and using the resulting DirEntry objects since they include info like whether its a link or directory directly from the syscall that walks the directory.  If an extra check for whether it exists as a file is necessary, that can be done through it and doing so will cache the stat info in the object for subsequent checks.  If it's not too much pain, using that and passing the direntry to the visitor could make a lot of operations faster in an interface like this.

Reply:
Oh, somehow completely missed that, that has the same functionality indeed. Thanks! Do you know if it's possible to get a deterministic order out of the entries? The tests assume alphabetic ordering now.

Reply:
Hm, `os.scandir` is a Python 3 thing. We can still define this method for Python 2 and then do something smarter for Python 3?

Reply:
If you mean with `walk()` itself, then It is with a roundabout way, as long as it's set for `topdown` mode, you can modify the `dirnames` list (modify in-place) however you want and it will walk in that order.  Here's the doc blurb:

> When topdown is True, the caller can modify the dirnames list in-place (perhaps using [del](dfile:///Users/scogland1/Library/Application%20Support/Dash/Versioned%20DocSets/Python%203%20-%20DHDocsetDownloader/3-9-5/Python%203.docset/Contents/Resources/Documents/doc/reference/simple_stmts.html#del) or slice assignment), and [walk()](dfile:///Users/scogland1/Library/Application%20Support/Dash/Versioned%20DocSets/Python%203%20-%20DHDocsetDownloader/3-9-5/Python%203.docset/Contents/Resources/Documents/doc/library/os.html#os.walk) will only recurse into the subdirectories whose names remain in dirnames; this can be used to prune the search, impose a specific order of visiting, or even to inform [walk()](dfile:///Users/scogland1/Library/Application%20Support/Dash/Versioned%20DocSets/Python%203%20-%20DHDocsetDownloader/3-9-5/Python%203.docset/Contents/Resources/Documents/doc/library/os.html#os.walk) about directories the caller creates or renames before it resumes [walk()](dfile:///Users/scogland1/Library/Application%20Support/Dash/Versioned%20DocSets/Python%203%20-%20DHDocsetDownloader/3-9-5/Python%203.docset/Contents/Resources/Documents/doc/library/os.html#os.walk) again. Modifying dirnames when topdown is False has no effect on the behavior of the walk, because in bottom-up mode the directories in dirnames are generated before dirpath itself is generated.

Since you want info about each entry though, and walk throws it away to provide path names, you really might want to keep this implementation but swapping `listdir` for `scandir` like walk uses internally instead to iterate over the more useful DirEntry objects since that will let you completely skip calling stat in several common cases.  Specifically the check for whether a target is a link, file, directory, block device or whatever are already available from the directory entry pulled from the readdir under the covers, the actual structure is here: https://www.gnu.org/software/libc/manual/html_node/Directory-Entries.html

A nice side-benefit is that if for some reason on a given target platform it isn't supported, the DirEntry wrapper will transparently add on the stat call and cache the result to make other checks on that object cheaper.  It's likely to make things downstream cheaper if we can manage to go that way.

Reply:
I've added a scandir default for Python 3.5